src/main/java/com/es/phoneshop/model/cart/DefaultCartService.java

```diff
@@ -0,0 +1,56 @@
+package com.es.phoneshop.model.cart;
+
+import com.es.phoneshop.exception.OutOfStockException;
+import com.es.phoneshop.model.product.ArrayListProductDao;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.product.ProductDao;
+import jakarta.servlet.http.HttpServletRequest;
+
+public class DefaultCartService implements CartService {
+
+    private static final String CART_SESSION_ATTRIBUTE = DefaultCartService.class.getName() + ".cart";
+    private static DefaultCartService instance;
+    private ProductDao productDao;
+
+    private DefaultCartService() {
+        productDao = ArrayListProductDao.getInstance();
+    }
+
+    public static synchronized DefaultCartService getInstance() {
+        if (instance == null) {
+            instance = new DefaultCartService();
+        }
+        return instance;
+    }
+
+    @Override
+    public synchronized Cart getCart(HttpServletRequest request) {
+        Cart cart = (Cart) request.getSession().getAttribute(CART_SESSION_ATTRIBUTE);
+        if (cart == null) {
+            request.getSession().setAttribute(CART_SESSION_ATTRIBUTE, cart = new Cart());
+        }
+        return cart;
+    }
+
+    @Override
+    public synchronized void add(Cart cart, Long productId, int quantity) throws OutOfStockException {
+        Product product = productDao.getProduct(productId);
+        if (product.getStock() < quantity) {
+            throw new OutOfStockException(product, quantity, product.getStock());
+        }
+
+        boolean itemNotExists = true;
+
+        for (CartItem cartItem : cart.getItems()) {
+            if (productId.equals(cartItem.getProduct().getId())) {
+                int previousQuantity = cartItem.getQuantity();
+                cartItem.setQuantity(previousQuantity + quantity);
```

Comment:
What if new quantity is bigger than stock?

Reply:
Added